File: crates/vexil-runtime/src/bit_writer.rs
pub struct BitWriter {
    buf: Vec<u8>,
    current_byte: u8,
    bit_offset: u8,
}

impl BitWriter {
    pub fn new() -> Self {
        Self {
            buf: Vec::new(),
            current_byte: 0,
            bit_offset: 0,
        }
    }

    /// Internal: align to a byte boundary without the "empty = zero byte" rule.
    /// Used before multi-byte writes to ensure alignment.
    fn align(&mut self) {
        if self.bit_offset > 0 {
            self.buf.push(self.current_byte);
            self.current_byte = 0;
            self.bit_offset = 0;
        }
    }

    /// Write `count` bits from `value`, LSB first.
    pub fn write_bits(&mut self, value: u64, count: u8) {
        let mut v = value;
        for _ in 0..count {
            let bit = (v & 1) as u8;
            self.current_byte |= bit << self.bit_offset;
            self.bit_offset += 1;
            if self.bit_offset == 8 {
                self.buf.push(self.current_byte);
                self.current_byte = 0;
                self.bit_offset = 0;
            }
            v >>= 1;
        }
    }

    /// Write a single boolean as 1 bit.
    pub fn write_bool(&mut self, v: bool) {
        self.write_bits(v as u64, 1);
    }

    /// Flush any partial byte to the buffer.
    ///
    /// Special case per spec §4.1: if nothing has been written at all
    /// (bit_offset == 0 AND buf is empty), push a zero byte anyway.
    /// If bit_offset == 0 and buf is non-empty, this is a no-op.
    pub fn flush_to_byte_boundary(&mut self) {
        if self.bit_offset == 0 {
            if self.buf.is_empty() {
                self.buf.push(0x00);
            }
            // else: already aligned and something was written — no-op
        } else {
            self.buf.push(self.current_byte);
            self.current_byte = 0;
            self.bit_offset = 0;
        }
    }

    pub fn write_u8(&mut self, v: u8) {
        self.align();
        self.buf.push(v);
    }

    pub fn write_u16(&mut self, v: u16) {
        self.align();
        self.buf.extend_from_slice(&v.to_le_bytes());
    }

    pub fn write_u32(&mut self, v: u32) {
        self.align();
        self.buf.extend_from_slice(&v.to_le_bytes());
    }

    pub fn write_u64(&mut self, v: u64) {
        self.align();
        self.buf.extend_from_slice(&v.to_le_bytes());
    }

    pub fn write_i8(&mut self, v: i8) {
        self.align();
        self.buf.extend_from_slice(&v.to_le_bytes());
    }

    pub fn write_i16(&mut self, v: i16) {
        self.align();
        self.buf.extend_from_slice(&v.to_le_bytes());
    }

    pub fn write_i32(&mut self, v: i32) {
        self.align();
        self.buf.extend_from_slice(&v.to_le_bytes());
    }

    pub fn write_i64(&mut self, v: i64) {
        self.align();
        self.buf.extend_from_slice(&v.to_le_bytes());
    }

    /// Write an f32, canonicalizing NaN to 0x7FC00000.
    pub fn write_f32(&mut self, v: f32) {
        self.align();
        let bits: u32 = if v.is_nan() {
            0x7FC00000u32
        } else {
            v.to_bits()
        };
        self.buf.extend_from_slice(&bits.to_le_bytes());
    }

    /// Write an f64, canonicalizing NaN to 0x7FF8000000000000.
    pub fn write_f64(&mut self, v: f64) {
        self.align();
        let bits: u64 = if v.is_nan() {
            0x7FF8000000000000u64
        } else {
            v.to_bits()
        };
        self.buf.extend_from_slice(&bits.to_le_bytes());
    }

    /// Write a LEB128-encoded unsigned integer.
    pub fn write_leb128(&mut self, v: u64) {
        self.align();
        crate::leb128::encode(&mut self.buf, v);
    }

    /// Write a ZigZag + LEB128 encoded signed integer.
    pub fn write_zigzag(&mut self, v: i64, type_bits: u8) {
        let encoded = crate::zigzag::zigzag_encode(v, type_bits);
        self.write_leb128(encoded);
    }

    /// Write a UTF-8 string with a LEB128 length prefix.
    pub fn write_string(&mut self, s: &str) {
        self.align();
        crate::leb128::encode(&mut self.buf, s.len() as u64);
        self.buf.extend_from_slice(s.as_bytes());
    }

    /// Write a byte slice with a LEB128 length prefix.
    pub fn write_bytes(&mut self, data: &[u8]) {
        self.align();
        crate::leb128::encode(&mut self.buf, data.len() as u64);
        self.buf.extend_from_slice(data);
    }

    /// Write raw bytes with no length prefix.
    pub fn write_raw_bytes(&mut self, data: &[u8]) {
        self.align();
        self.buf.extend_from_slice(data);
    }

    /// Flush any partial byte and return the finished buffer.
    pub fn finish(mut self) -> Vec<u8> {
        self.flush_to_byte_boundary();
        self.buf
    }
}

impl Default for BitWriter {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn write_single_bit_true() {
        let mut w = BitWriter::new();
        w.write_bool(true);
        assert_eq!(w.finish(), [0x01]);
    }

    #[test]
    fn write_single_bit_false() {
        let mut w = BitWriter::new();
        w.write_bool(false);
        assert_eq!(w.finish(), [0x00]);
    }

    #[test]
    fn write_bits_lsb_first() {
        let mut w = BitWriter::new();
        w.write_bits(5, 3); // 101
        w.write_bits(19, 5); // 10011
                             // LSB-first: byte = 10011_101 = 0x9D
        assert_eq!(w.finish(), [0x9D]);
    }

    #[test]
    fn write_bits_cross_byte_boundary() {
        let mut w = BitWriter::new();
        w.write_bits(5, 3);
        w.write_bits(19, 5);
        w.write_bits(42, 6); // 101010
                             // Byte 0: 0x9D, Byte 1: 00_101010 = 0x2A
        assert_eq!(w.finish(), [0x9D, 0x2A]);
    }

    #[test]
    fn flush_to_byte_boundary_pads_zeros() {
        let mut w = BitWriter::new();
        w.write_bits(0b101, 3);
        w.flush_to_byte_boundary();
        w.write_bits(0xFF, 8);
        assert_eq!(w.finish(), [0x05, 0xFF]);
    }

    #[test]
    fn write_u8_flushes_first() {
        let mut w = BitWriter::new();
        w.write_bool(true);
        w.write_u8(0xAB);
        assert_eq!(w.finish(), [0x01, 0xAB]);
    }

    #[test]
    fn write_u16_le() {
        let mut w = BitWriter::new();
        w.write_u16(0x0102);
        assert_eq!(w.finish(), [0x02, 0x01]);
    }

    #[test]
    fn write_u32_le() {
        let mut w = BitWriter::new();
        w.write_u32(0x01020304);
        assert_eq!(w.finish(), [0x04, 0x03, 0x02, 0x01]);
    }

    #[test]
    fn write_i16_negative() {
        let mut w = BitWriter::new();
        w.write_i16(-1);
        assert_eq!(w.finish(), [0xFF, 0xFF]);
    }

    #[test]
    fn write_f32_nan_canonicalized() {
        let mut w = BitWriter::new();
        w.write_f32(f32::NAN);
        assert_eq!(w.finish(), [0x00, 0x00, 0xC0, 0x7F]);
    }

    #[test]
    fn write_f64_nan_canonicalized() {
        let mut w = BitWriter::new();
        w.write_f64(f64::NAN);
        assert_eq!(w.finish(), 0x7FF8000000000000u64.to_le_bytes());
    }

    #[test]
    fn write_f32_negative_zero_preserved() {
        let mut w = BitWriter::new();
        w.write_f32(-0.0f32);
        let buf = w.finish();
        assert_eq!(buf, (-0.0f32).to_le_bytes());
        assert_ne!(buf, 0.0f32.to_le_bytes());
    }

    #[test]
    fn write_leb128_test() {
        let mut w = BitWriter::new();
        w.write_leb128(300);
        assert_eq!(w.finish(), [0xAC, 0x02]);
    }

    #[test]
    fn write_zigzag_neg1() {
        let mut w = BitWriter::new();
        w.write_zigzag(-1, 64);
        assert_eq!(w.finish(), [0x01]);
    }

    #[test]
    fn write_string_test() {
        let mut w = BitWriter::new();
        w.write_string("hi");
        assert_eq!(w.finish(), [0x02, 0x68, 0x69]);
    }

    #[test]
    fn write_bytes_test() {
        let mut w = BitWriter::new();
        w.write_bytes(&[0xDE, 0xAD]);
        assert_eq!(w.finish(), [0x02, 0xDE, 0xAD]);
    }

    #[test]
    fn write_raw_bytes_test() {
        let mut w = BitWriter::new();
        w.write_raw_bytes(&[0xCA, 0xFE]);
        assert_eq!(w.finish(), [0xCA, 0xFE]);
    }

    #[test]
    fn empty_flush_produces_zero_byte() {
        let mut w = BitWriter::new();
        w.flush_to_byte_boundary();
        assert_eq!(w.finish(), [0x00]);
    }
}

File: packages/runtime-ts/src/bit-writer.ts
const MAX_RECURSION_DEPTH = 64;

const textEncoder = new TextEncoder();

/**
 * A byte-buffer builder that packs fields LSB-first at the bit level.
 *
 * Created with `new BitWriter()`, written to with `write*` methods, and
 * finalized with `finish()` which flushes any partial byte and returns
 * the completed buffer.
 *
 * Sub-byte fields are accumulated in a single byte; once 8 bits are filled
 * the byte is flushed. Multi-byte writes (e.g. `writeU16`) first align to
 * a byte boundary, then append little-endian bytes directly.
 */
export class BitWriter {
  private buf: number[] = [];
  private currentByte = 0;
  private bitOffset = 0;
  private recursionDepth = 0;

  /**
   * Internal: align to a byte boundary without the "empty = zero byte" rule.
   * Used before multi-byte writes to ensure alignment.
   */
  private align(): void {
    if (this.bitOffset > 0) {
      this.buf.push(this.currentByte);
      this.currentByte = 0;
      this.bitOffset = 0;
    }
  }

  /**
   * Write `count` bits from `value`, LSB first.
   */
  writeBits(value: number, count: number): void {
    let v = value;
    for (let i = 0; i < count; i++) {
      const bit = v & 1;
      this.currentByte |= bit << this.bitOffset;
      this.bitOffset++;
      if (this.bitOffset === 8) {
        this.buf.push(this.currentByte);
        this.currentByte = 0;
        this.bitOffset = 0;
      }
      v >>>= 1;
    }
  }

  /**
   * Write a single boolean as 1 bit.
   */
  writeBool(v: boolean): void {
    this.writeBits(v ? 1 : 0, 1);
  }

  /**
   * Flush any partial byte to the buffer.
   *
   * Special case per spec section 4.1: if nothing has been written at all
   * (bitOffset == 0 AND buf is empty), push a zero byte anyway.
   * If bitOffset == 0 and buf is non-empty, this is a no-op.
   */
  flushToByteBoundary(): void {
    if (this.bitOffset === 0) {
      if (this.buf.length === 0) {
        this.buf.push(0x00);
      }
      // else: already aligned and something was written - no-op
    } else {
      this.buf.push(this.currentByte);
      this.currentByte = 0;
      this.bitOffset = 0;
    }
  }

  /**
   * Write a u8, aligning to a byte boundary first.
   */
  writeU8(v: number): void {
    this.align();
    this.buf.push(v & 0xff);
  }

  /**
   * Write a u16 in little-endian byte order, aligning first.
   */
  writeU16(v: number): void {
    this.align();
    this.buf.push(v & 0xff);
    this.buf.push((v >>> 8) & 0xff);
  }

  /**
   * Write a u32 in little-endian byte order, aligning first.
   */
  writeU32(v: number): void {
    this.align();
    this.buf.push(v & 0xff);
    this.buf.push((v >>> 8) & 0xff);
    this.buf.push((v >>> 16) & 0xff);
    this.buf.push((v >>> 24) & 0xff);
  }

  /**
   * Write a u64 in little-endian byte order, aligning first.
   */
  writeU64(v: bigint): void {
    this.align();
    const mask = BigInt(0xff);
    for (let i = 0; i < 8; i++) {
      this.buf.push(Number((v >> BigInt(i * 8)) & mask));
    }
  }

  /**
   * Write an i8, aligning to a byte boundary first.
   */
  writeI8(v: number): void {
    this.align();
    this.buf.push(v & 0xff);
  }

  /**
   * Write an i16 in little-endian byte order, aligning first.
   */
  writeI16(v: number): void {
    this.align();
    // Use DataView for correct two's complement encoding
    const ab = new ArrayBuffer(2);
    new DataView(ab).setInt16(0, v, true);
    const bytes = new Uint8Array(ab);
    this.buf.push(bytes[0], bytes[1]);
  }

  /**
   * Write an i32 in little-endian byte order, aligning first.
   */
  writeI32(v: number): void {
    this.align();
    const ab = new ArrayBuffer(4);
    new DataView(ab).setInt32(0, v, true);
    const bytes = new Uint8Array(ab);
    this.buf.push(bytes[0], bytes[1], bytes[2], bytes[3]);
  }

  /**
   * Write an i64 in little-endian byte order, aligning first.
   */
  writeI64(v: bigint): void {
    this.align();
    const ab = new ArrayBuffer(8);
    new DataView(ab).setBigInt64(0, v, true);
    const bytes = new Uint8Array(ab);
    for (let i = 0; i < 8; i++) {
      this.buf.push(bytes[i]);
    }
  }

  /**
   * Write an f32, canonicalizing NaN to 0x7FC00000.
   */
  writeF32(v: number): void {
    this.align();
    const ab = new ArrayBuffer(4);
    const dv = new DataView(ab);
    if (Number.isNaN(v)) {
      dv.setUint32(0, 0x7fc00000, true);
    } else {
      dv.setFloat32(0, v, true);
    }
    const bytes = new Uint8Array(ab);
    this.buf.push(bytes[0], bytes[1], bytes[2], bytes[3]);
  }

  /**
   * Write an f64, canonicalizing NaN to 0x7FF8000000000000.
   */
  writeF64(v: number): void {
    this.align();
    const ab = new ArrayBuffer(8);
    const dv = new DataView(ab);
    if (Number.isNaN(v)) {
      // Canonical qNaN: 0x7FF8000000000000
      dv.setUint32(0, 0x00000000, true); // low 4 bytes LE
      dv.setUint32(4, 0x7ff80000, true); // high 4 bytes LE
    } else {
      dv.setFloat64(0, v, true);
    }
    const bytes = new Uint8Array(ab);
    for (let i = 0; i < 8; i++) {
      this.buf.push(bytes[i]);
    }
  }

  /**
   * Write a LEB128-encoded unsigned integer.
   */
  writeLeb128(value: number): void {
    this.align();
    let v = value;
    do {
      let byte = v & 0x7f;
      v >>>= 7;
      if (v !== 0) {
        byte |= 0x80;
      }
      this.buf.push(byte);
    } while (v !== 0);
  }

  /**
   * Write a LEB128-encoded unsigned 64-bit integer (bigint).
   */
  writeLeb12864(value: bigint): void {
    this.align();
    let v = value < 0n ? value + (1n << 64n) : value; // treat as unsigned
    do {
      let byte = Number(v & 0x7fn);
      v >>= 7n;
      if (v !== 0n) {
        byte |= 0x80;
      }
      this.buf.push(byte);
    } while (v !== 0n);
  }

  /**
   * Write a ZigZag + LEB128 encoded signed integer (up to 32-bit).
   */
  writeZigZag(value: number, typeBits: number): void {
    // ZigZag encode: (n << 1) ^ (n >> (bits - 1))
    const zigzag = (value << 1) ^ (value >> (typeBits - 1));
    this.writeLeb128(zigzag >>> 0); // treat as unsigned
  }

  /**
   * Write a ZigZag + LEB128 encoded signed 64-bit integer (bigint).
   */
  writeZigZag64(value: bigint): void {
    const zigzag = (value << 1n) ^ (value >> 63n);
    this.writeLeb12864(zigzag < 0n ? zigzag + (1n << 64n) : zigzag);
  }

  /**
   * Write a UTF-8 string with a LEB128 length prefix.
   */
  writeString(s: string): void {
    this.align();
    const encoded = textEncoder.encode(s);
    this.writeLeb128Internal(encoded.length);
    for (let i = 0; i < encoded.length; i++) {
      this.buf.push(encoded[i]);
    }
  }

  /**
   * Write a byte array with a LEB128 length prefix.
   */
  writeBytes(data: Uint8Array): void {
    this.align();
    this.writeLeb128Internal(data.length);
    for (let i = 0; i < data.length; i++) {
      this.buf.push(data[i]);
    }
  }

  /**
   * Write raw bytes with no length prefix, aligning first.
   */
  writeRawBytes(data: Uint8Array | number[]): void {
    this.align();
    for (let i = 0; i < data.length; i++) {
      this.buf.push(data[i]);
    }
  }

  /**
   * Increment recursion depth; throws if limit exceeded.
   */
  enterNested(): void {
    this.recursionDepth++;
    if (this.recursionDepth > MAX_RECURSION_DEPTH) {
      throw new Error('Recursion limit exceeded');
    }
  }

  /**
   * Decrement recursion depth.
   */
  leaveNested(): void {
    if (this.recursionDepth > 0) {
      this.recursionDepth--;
    }
  }

  /**
   * Flush any partial byte and return the finished buffer.
   */
  finish(): Uint8Array {
    this.flushToByteBoundary();
    return new Uint8Array(this.buf);
  }

  /**
   * Internal LEB128 encode that does NOT call align() — used by writeString/writeBytes
   * which have already aligned.
   */
  private writeLeb128Internal(value: number): void {
    let v = value;
    do {
      let byte = v & 0x7f;
      v >>>= 7;
      if (v !== 0) {
        byte |= 0x80;
      }
      this.buf.push(byte);
    } while (v !== 0);
  }
}
